### DISCUSSION TITLE
How to save pokemon

### DISCUSSION THREAD

**Discussion by bisigoof** [deltabeard/Peanut-GB#112] [Q&A] [answered]
Im sorry if im asking something dumb, i know nothing of coding and i just want to get pokemon (that i can save) on my calculator. Is it possible, if yes how???? (also if ur gonna say something with coding please make it easy for me to understand i dont know coding)

**Comment by deltabeard** [comment]
What calculator is it and where did you download the program from?

**Reply by bisigoof** [reply]
I got the file from this page: https://github.com/nwagyu/peanutgb
The calculator is numworks, I got pokemon on it but i cant save the game which makes it practically useless

**Reply by deltabeard** [reply, marked as answer]
That app doesn't have save support unfortunately. You can open an issue on that repository if you want, but I can't provide support for it as I don't have that calculator and don't know anything about the platform.

**Reply by Yaya-Cout** [reply]
**TLDR: If you want save support on your NumWorks calculator, use the version available here: https://yaya-cout.github.io/Nwagyu/guide/apps/peanut-gb.html**

Hi!
I'm the current maintainer of the Peanut-GB app for NumWorks. First of all, thanks you for this library, it would have probably been impossible to run GB (and GBC) on the NumWorks without it. Some specs if you are interested STM32F730 MCU, with a Cortex-M7 192 MHz MCU, 256 KB of RAM (+ 16 KB of ITCM which is used as data RAM in some cases, but with around 80 KB really usable as the OS already need quite a lot of memory), 8 MB of flash (usually not written during runtime). The specs are quite similar to RP2040.

Here's a bit of (simplified) history to explain a bit more (especially why the name hasn't been changed), feel free to skip to the end of the post if you don't want to read it:
In the beginning, the NumWorks calculator was open-source (well, the license can't be qualified as such, but way more open than Casio and TI, the license was CC BY-NC-SA 4.0), named Epsilon. Forks of the firmware were created (Omega and Delta), which restored symbolic calculator and added an external apps framework (which I usually refer as `extapp` API), at first to integrate Giac/xCAS a CAS engine released as GPL 3.0, so it couldn't be directly linked with the OS. After porting it as an external app (name KhiCAS), other apps were created: a Periodic table and emulators ports: Nofrendo for the NES, with savestate support, and Peanut-GB was chosen for Game-Boy.
The apps were distributed through a website with a selector then directly installed to the NumWorks, so Peanut-GB name made sense, at it was mostly just a wrapper, in a repo merged with a lot of other apps, so the name didn't get referenced by search engines.

After some time, NumWorks changed their mind, decided to go back on user freedom and changed their license (they removed it, so we can't do anything with the new code except within the copyright exceptions, which are in France for interoperability, private use and terms granted by GitHub). They did so to be able to sell in some countries (Portugal and Netherlands). They also locked the bootloader to prevent anyone from flashing a custom firmware (you can still flash a custom userland, with severe limitations like not being able to use it during exams, and the code to do so is under strong copyright). At the same time, to keep their reputation of being "open", they created their own external apps framework (named `nwa`), with some differences compared to the `extapp` framework from Omega, blocked in exam mode (blocking KhiCAS, the app allowing to do advanced maths, even if it's allowed in exam in France). Another big limitation was the absence of any storage access API.

At the same time, the Omega project (main custom firmware) was archived and the Upsilon fork was created to continue it (I'm one of the maintainers of this fork, but it's loosing speed…). We got in a situation where firmwares diverged.

At some point, ScarlettSpell, a person not known to the community before, decided to port all the apps from Upsilon (in the `extapp` API) to the official firmware (`nwa` API). Features impossible to implement using the API were simply removed, such as save support in Peanut-GB. This was the moment Peanut-GB got it's own repo, and even a page on the `nwagyu.com` website. The name was kept as it was a port of the Upsilon app. After doing the initial ports, this person stopped sending any message in the Discord server (1 month and 2 weeks after joining), and one year later, the website went down as the domain name wasn't renewed.

When the website went down, I quickly recreated it using VuePress to preserve the informations from this website. I also forked the apps and added some features. As I worked on Upsilon, I knew the low-level file system structure and wrote an API for it, which isn't perfect, and still have some corruption issues as of today I can't reproduce (code is here: https://framagit.org/Yaya.Cout/numworks-extapp-storage/-/blob/master/src/storage.c?ref_type=heads).
After I wrote this API, which I designed to be almost source-compatible with `extapp` API syntax, I implemented save support in Peanut-GB in the `nwa` version, so anyone with a NumWorks calculator is now able to use Peanut-GB with save support (port code is here: https://codeberg.org/Yaya-Cout/peanutgb). I also updated the existing GBC port (which was based on PR code back then with all the new features).

And now, we are mostly stuck with the name Peanut-GB for the NumWorks app, which is now referenced by most tutorials. I don't know if you consider that as a problem, and if you have any idea how to fix this.

I'm still making improvements to the port (I don't have access to my calculator since 2 months so I stopped working on it), for example to the scaling algorithm (I tried to do it using subpixel rendering on the 320×240 screen to get a round number of pixels to scale and better results than having doubled lines then not doubled lines on the render, which is really bad), to the frame skipping code (switching dynamically between 60FPS/30FPS/15FPS to get the maximum framerate possible while still keeping up with realtime, interlacing was looking really bad on the screen configured at 40 FPS, even with 60 FPS output from Peanut-GB), trying to improve performance (compiler flags and frontend code optimisation, finding better tradeoff). We also modified a small part of Peanut-GB to be able to allocate buffers manually (replacing `uint8_t vram[VRAM_SIZE];` by `uint8_t* vram;`, and setting it using `gb.wram = malloc(WRAM_SIZE)` just before `gb_init`)

You can add the Codeberg link to the "Projects Using Peanut-GB" list, if you want.

Again, thanks for this wonderful library.
